controllers/auth_config_status_updater.go

```diff
@@ -41,41 +46,152 @@ func (u *AuthConfigStatusUpdater) Reconcile(ctx context.Context, req ctrl.Reques
-	authConfig.Status.NumAuthorizationPolicies = int64(len(authConfig.Spec.Authorization))
-	authConfig.Status.NumResponseItems = int64(len(authConfig.Spec.Response))
+	var conditionsChanged, summaryChanged bool
 
-	issuingWristbands := false
-	for _, responseConfig := range authConfig.Spec.Response {
-		if responseConfig.GetType() == api.ResponseWristband {
-			issuingWristbands = true
-			break
-		}
+	authConfig.Status.Conditions, conditionsChanged = buildStatusConditions(authConfig.Status.Conditions, condition)
+	authConfig.Status.Summary, summaryChanged = buildStatusSummary(authConfig, linked)
+
+	if !authConfig.Status.Ready() {
+		err = fmt.Errorf("resource not ready")
 	}
-	authConfig.Status.FestivalWristbandEnabled = issuingWristbands
 
-	return u.Status().Update(ctx, authConfig)
+	if changed := conditionsChanged || summaryChanged; !changed {
+		logger.V(1).Info("resource status did not change")
+		return // to save an update request
+	}
+
+	logger.V(1).Info("resource status changed", "authconfig/status", authConfig.Status)
+
+	if updateErr := u.Status().Update(ctx, authConfig); updateErr != nil {
+		logger.Error(updateErr, "failed to update the resource")
+		err = updateErr
+	}
+
+	logger.Info("resource status updated")
+
+	return
+}
+
+func (u *AuthConfigStatusUpdater) partitionHostsByStatus(cacheId string, authConfig *api.AuthConfig) ([]string, []string) {
+	type obj struct{}
+	hosts := make(map[string]obj)
+	for _, host := range u.Cache.FindKeys(cacheId) {
+		hosts[host] = obj{}
+	}
+	linked := []string{}
+	loose := []string{}
+	for _, host := range authConfig.Spec.Hosts {
+		if _, ok := hosts[host]; ok {
+			linked = append(linked, host)
+		} else {
+			loose = append(loose, host)
+		}
+	}
+	return linked, loose
 }
 
 func (u *AuthConfigStatusUpdater) SetupWithManager(mgr ctrl.Manager) error {
 	return ctrl.NewControllerManagedBy(mgr).
 		For(&api.AuthConfig{}, builder.WithPredicates(LabelSelectorPredicate(u.LabelSelector))).
 		Complete(u)
 }
+
+func statusReady() api.Condition {
+	return api.Condition{
+		Type:   api.StatusConditionReady,
+		Status: k8score.ConditionTrue,
+		Reason: api.StatusReasonReconciled,
+	}
+}
+
+func statusNotReady(reason, message string) api.Condition {
+	return api.Condition{
+		Type:    api.StatusConditionReady,
+		Status:  k8score.ConditionFalse,
+		Reason:  reason,
+		Message: message,
+	}
+}
+
+func buildStatusConditions(current []api.Condition, new api.Condition) ([]api.Condition, bool) {
+	new.LastTransitionTime = metav1.Now()
+
+	if current == nil {
+		return []api.Condition{new}, true
+	}
+
+	for i, curr := range current {
+		if curr.Type == new.Type {
+			if curr.Status == new.Status {
+				if curr.Reason == new.Reason && curr.Message == new.Message {
+					return current, false
+				}
+
+				new.LastTransitionTime = curr.LastTransitionTime
+			}
+
+			res := make([]api.Condition, len(current))
+			copy(res, current)
+			res[i] = new
+			return res, true
```

Comment:
Could it be the case that a new condition that has same type, different status would force us return an "early" array of conditions leaving behind other conditions to check?

Reply:
Not really, because `buildStatusConditions` focuses on one "new" condition at a time. The goal is to get that one condition either added or updated in the existing array of conditions. Once that one is found, there's no need to keep checking for others.

We could be invoking `buildStatusConditions` multiple times, accumulating changes, one by one, into a variable `conditions` and then set `authConfig.Status.Conditions = conditions`. Right now, the whole status update only adds/updates one condition; because of that, we set `authConfig.Status.Conditions, ... = buildStatusConditions(...)` directly.